databroker/broker.py

```diff
@@ -880,3 +902,161 @@ def _munge_time(t, timezone):
+                        filled={k: True for k in external_writers.keys()})
+
+                elif name == 'stop':
+                    for data_key, writer in list(writers.items()):
+                        writer.close()
+                        writers.pop(data_key)
+
+                # The mutated fs_doc is inserted into metadatastore.
+                fs_doc.pop('filled', None)
+                fs_doc.pop('_name', None)
+                db.mds.insert(name, fs_doc)
+
+                # The pristine doc is yielded.
+                yield name, doc
+
+        return wrapped_f
+
+    return wrap
+
+
+def event_map(stream_name, data_keys, provenance):
+    """
+    Map a function onto each event in a stream.
+
+    Parameters
+    ----------
+    stream_name : string
+        e.g., 'primary' or 'baseline'
+    data_keys : dict
+        key(s) in event['data'] to apply function to (e.g., 'image') mapped to
+        a dict with any updates to the data key value
+        (e.g., {'shape': [10, 10]}). In the simple case where the shape,
+        datatype, etc. are unchanged, the dict is just empty:
+        ``{'image': {}}`` and the original metadata passed through.
+    proveance : dict
+        metadata about this operation
+    """
+    def outer(f):
+        def inner(stream):
+            run_start_uid = None
+            descriptor_uid = None
+            for name, doc in stream:
+                if name == 'start':
+                    run_start_uid = str(uuid.uuid4())
+                    new_start_doc = dict(uid=run_start_uid,
+                                         time=time.time(),
+                                         parents=[doc['uid']],
+                                         provenance=provenance)
+                    yield 'start', new_start_doc
+
+                elif name == 'descriptor' and doc.get('name') == stream_name:
+                    if run_start_uid is None:
+                        raise RuntimeError("Received EventDescriptor before "
+                                           "RunStart.")
+                    descriptor_uid = doc_or_uid_to_uid(doc)
+                    new_data_keys = dict(doc['data_keys'])
+                    for k, v in new_data_keys.items():
+                        new_data_keys[k].update(v)
+                    new_descriptor = dict(uid=str(uuid.uuid4()),
+                                          time=time.time(),
+                                          run_start=run_start_uid,
+                                          data_keys=new_data_keys,
+                                          name=stream_name)
+                    yield 'descriptor', new_descriptor
+
+                elif (name == 'event' and
+                      doc_or_uid_to_uid(doc['descriptor']) == descriptor_uid):
+                    if run_start_uid is None:
+                        raise RuntimeError("Received Event before RunStart.")
+                    try:
+                        new_event = dict(doc)
```

Comment:
I agree, that's a good idea. I don't think our analysis pipelines so far are well defined yet on our end (CMS), and they can be quite complicated (lots of inputs, lots of outputs, sometimes input is db data, sometimes something else) so I think we may take the more argument based approach and leave this aside. For ex:
```python 
def process(*args, **kwargs):
    for arg in args:
        if isinstance(arg, Header):
            # do some reading replace arg with what header has
    # now run function stuff
```
etc. (of course this would be done with decorators acting on functions) The storing I imagine would just be one document (since an event is always just one function in my mind, even a 'stream' doesn't make sense since outside functions in another stream may eventually depend on an intermediate step in another 'stream'. It's more a huge interconnected graph.).
Later on, we can use our implementation to decide whether `event_map` (and permutations) are useful and build them in. That's our take for now, just to give context. We can hold off discussion on this for later, @danielballan is helping moderate our approaches.

I definitely see the use for this routine, it's very quick for incoming data. I think it is suitable. My only concern is how to save the documents. But perhaps if it's a batch processed stream, it should remain as a stream (and not something other, like an 'analysis' document). Sorry for the long comment. Might be better suited in an eventual discussion in person.